### ISSUE TITLE
List of bean by types autowiring in Kotlin AOT not working since 6.2.0

### ISSUE THREAD

**Issue by artemptushkin** [spring-projects/spring-framework#34122]
I have Spring Boot 3.4.0 project that uses Spring Boot AOT plugin `org.springframework.boot.aot` so the Java runtime runs with `-Dspring.aot.enabled=true`. I don't use Spring GraalVM native, so it's only AOT.

I have the following Kotlin code:

```kt
@Configuration
class TelegramCommandsConfiguration(
    private val commandsBeans: List<TelegramCommand>,
    private val callbackBeans: List<TelegramCallbackHandler>,
) {

    @Bean
    fun commands(): Map<String, TelegramCommand> = commandsBeans.associateBy { it.name() } // returns 0

    @Bean
    fun callbacks(): Map<String, TelegramCallbackHandler> = callbackBeans.associateBy { it.name() } // returns 0
}
```

Where:
* `TelegramCommand` is an interface
* there are N beans of it, some defined with `@Bean` and some defined with `@Component`
* Being on SB 3.3.6 and Spring 6.1.15 all are fetched as expected with the `List` access.

<img width="968" alt="image" src="https://github.com/user-attachments/assets/9e52591b-e69f-4908-a291-68d4e81d4142" />

But since I updated SB to 3.4.0 and made no other changes to the list variables, I fetched **0 beans** by the interface `TelegramCommand`.

<img width="736" alt="image" src="https://github.com/user-attachments/assets/310d0370-3197-4457-a2ff-3b1702546845" />

I read about the recent changes in [core container](https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-6.2-Release-Notes#core-container) but I can not find a clue about what I should do now.

Is this an issue?

----

I tried to run w/o AOT - all good the beans of `TelegramCommand` are there so it's something about AOT.

----

Workaround, explicit access:

```kt
@Bean
    fun commands(): Commands = Commands(
        applicationContext.getBeansOfType<TelegramCommand>(TelegramCommand::class.java)
            .entries
            .associateBy({ it.value.name() }, { it.value })
    )
````

**Comment by sdeleuze** [comment]
Could you please test again with Spring Boot 3.4.1 and if that still occur, share a reproducer as an archive attached to this issue or as a GitHub repository?

**Comment by artemptushkin** [comment]
@sdeleuze indeed it's fixed today recently released Spring Boot 3.4.1! Thank you
